xwiki-commons-core/xwiki-commons-xml/src/test/java/org/xwiki/xml/internal/html/DefaultHTMLCleanerTest.java

```diff
@@ -533,25 +535,27 @@ void controlCharacters() throws Exception
         htmlInput = "<p>&#8;</p>";
         document = this.cleaner.clean(new StringReader(htmlInput));
 
-        // HtmlCleaner currently doesn't handle properly unicode characters: asking it to recognize them
-        // involves that all entities will be escaped during the parsing and that's not what we want. So we
-        // keep them encoded.
+        // HtmlCleaner now handles unicode character entities. These are decoded
+        // if possible, and if they are control characters, replaced by whitespace
         // See https://sourceforge.net/p/htmlcleaner/bugs/221/
         textContent =
             document.getElementsByTagName("p").item(0).getTextContent();
-        assertEquals("&#8;", textContent);
-        assertHTML("<p>&#8;</p>", "&#8;");
+        assertEquals(" ", textContent);
+        assertHTML(" ", "&#8;");
 
         htmlInput = "<p foo=\"&#8;\">content</p>";
         document = this.cleaner.clean(new StringReader(htmlInput));
 
-        // HtmlCleaner currently doesn't handle properly unicode characters: asking it to recognize them
-        // involves that all entities will be escaped during the parsing and that's not what we want. So we
-        // keep them encoded.
+        // in attributes we actually want to keep (encoded) control characters
+        // but HTMLCleaner scraps these characters if it finds them in attributes
         textContent =
             document.getElementsByTagName("p").item(0).getAttributes().getNamedItem("foo").getTextContent();
-        assertEquals("&#8;", textContent);
-        assertHTML("<p foo=\"&#8;\">content</p>", "<p foo=\"&#8;\">content</p>");
+        // wanted behavior:
+        // assertEquals("&#8;", textContent);
+        // actual behavior
```

Comment:
I'm not sure I understand this. Are we testing for a behavior we don't want? that doesn't look good. Tests are the spec.

I think a better approach is to add a test with an ignore annotation and a description.

WDYT?

Reply:
Also is this a change from before the upgrade?

Reply:
Yes, this is a change introduced by the PR, by setting `defaultProperties.setRecognizeUnicodeChars(true)` for the HTML cleaner configuration.

Before the change, if some characters have embedded as entities (i.e. here `&#8;`) they have been inserted as such in the HTML and then "shown" in the browser as their character value (i.e. here a backspace).
 
However now they are first replaced by their actual, well, character (in this case as a single-byte character corresponding to `\b`, i.e backspace), because they need no entity encoding/escaping (only `&`, `<`, `>`, `"` and `'` do need it.).
In a second step the HTML-Cleaner detects those characters which are control characters and replaces them by white space (in text nodes) or removes them (in attribute text). I guess that is because the HTMLCleaner thinks control characters are not legit in HTML text or attributes nodes.

 Thus the attribute passed in as `foo="&#8;"`  is reduced to `foo=""`. in the previous version it was kept as `foo="&#8;"`. I kept the old behavior as  "wanted behavior" in the code as I am uncertain if this is wanted or not.

Note that line breaks are not subject to removal in text nodes. This is tested below in "doNotRemoveLFandCRInTextNodes"

Reply:
> However now they are first replaced by their actual, well, character (in this case as a single-byte character corresponding to \b, i.e backspace), because they need no entity encoding/escaping (only &, <, >, " and ' do need it.).

So that behaviour is checked first on previous part of the test, and I agree with it.

> In a second step the HTML-Cleaner detects those characters which are control characters and replaces them by white space (in text nodes) or removes them (in attribute text). I guess that is because the HTMLCleaner thinks control characters are not legit in HTML text or attributes nodes.

That second behaviour is a bit less clear: it would be interesting to make the same test before the HtmlCleaner upgrade with `\u0008` value in the attribute since AFAICS it should behave the same.

Reply:
I updated the comment with latest info from @mflorea